### ISSUE TITLE
Mark of Grace spawn plugin (Ardougne Rooftop course)

### ISSUE THREAD

**Issue by renatosj** [runelite/runelite#11485]
**Is your feature request related to a problem? Please describe.**
Im kinda of annoyed that i've "found out" that the spawn chance of marks of grace is not rng based, but its based on a time spent on the roof course...

**Describe the solution you'd like**
So i'd like to request a timer starting at the moment you climb the first obstacle, so i can see how much time i've spent at the roof; the timer would reset after the last xp drop (529) and start once again after the first xp drop/first obstacle.

**Describe alternatives you've considered**
A simple timer with a MoG picture (use the boosts timer as an example) would be just fine

**Additional context**
So basically, idk about the others agility rooftop courses, but for Ardougne, the chance to spawn a Mark of Grace (idk if the diary affects that) seems to be every ~2min spent while on the roof course, so assuming that is correct, the ammount of MoG p/h would be 30.

**Comment by Nightfirecat** [comment]
Without knowing exact mechanics, I don't think we're going to add this into the client.

**Comment by renatosj** [comment]
Is it possible for us to find out how does it work? Does simply asking a jmod would help/solve this issue?

**Comment by Just-Sin** [comment]
without knowing the exact mechanics, could this idea be changed to a timer since the last mark of grace has spawn? So for example when the current plugin turns the agility boxes red (detects mark of grace spawn), it starts a stopwatch like timer counting up and resets to 0 when the next mark is spawned (without looking at the source not sure if this would be a problem if you don't pick up marks, but should be fixable if so).

I personally believe that a mark can spawn randomly while completing the course obstacles, there's a 2 min timer (from my testing 2 minutes seems to be consistent with the ardy 4 diary) before another can spawn, but i believe you have to complete a lap (xp drop for last obstacle [maybe a fail but im 99 so can't test]) after the 2 minute timer has ended in order to roll for your next mark and the cycle start again.

If this theory is right a timer since last mark spawn (could be paired with an additional feature showing your tick perfect time shown (known value, ardy 76t (45.6s) https://oldschool.runescape.wiki/w/Rooftop_Agility_Courses) and average times for say your last 5 laps (obviously not everybody is tick perfect and afks) could let you know if you should wait at last obstacle for 5 seconds (say if timer >= 1:56s), or rush another lap (say if the timer is <= 1:05s).

It doesn't rely on knowing the exact mechanics, but with data collected from this the mechanic could likely be worked out slowly in the future, and I don't think it would be very hard to implement... i'll start a fork and try to show what i'm talking about - although i'm not java expert should be most copy and paste from the existing plugin, see what I can come up with

**Comment by Mrnice98** [comment]
I'm slightly confused what you are asking for, do you want a timer that will let you know how long its been since the last mark was added on the tile in Ardougne ? or is this a theory you have about how the marks of grace are rolled and this timer would aid in trying to get the max spawns per hour ?

**Comment by Just-Sin** [comment]
both! im pretty sure I understand the mechanics, and the timer will make it so I don't have to keep a stop watch going. Once I understand the mechanic more I can try to get confirmation from jmods and update plugin to reflect it

**Comment by Just-Sin** [comment]
if you're more familiar with the code, would it sense when the stack changes from mark of grace -> mark of grace (2)?

**Comment by Just-Sin** [comment]
I created a fork at https://github.com/Just-Sin/runelite and branch 'mog-spawn-optimization-helper' if you're thinking about editing anything

**Comment by Mrnice98** [comment]
if the marks are always in view I suppose it could be done. I'm not super familiar with the agility plugin I was just more confused about what was being asked. I think you would have more luck trying to create a plugin hub, plugin for something like this though as  this is a closed issue and until there is a definitive answer on how the mechanics work I don't think that will change.

**Comment by Just-Sin** [comment]
at the end of the day im just adding an option to add a timer, the rest is my speculation so I do think it belongs in the actual agility plugin
